File: Shreya Suresh/ads shre/rbt.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Node {
    int data;
    struct Node *left, *right, *parent;
    bool color;
} Node;

Node* newNode(int data) {
    Node* n = (Node*)malloc(sizeof(Node));
    n->data = data;
    n->left = n->right = n->parent = NULL;
    n->color = true;
    return n;
}

void leftRotate(Node **root, Node *x) {
    Node *y = x->right;
    x->right = y->left;
    if (y->left != NULL)
        y->left->parent = x;

    y->parent = x->parent;
    if (x->parent == NULL)
        *root = y;
    else if (x == x->parent->left)
        x->parent->left = y;
    else
        x->parent->right = y;

    y->left = x;
    x->parent = y;
}

void rightRotate(Node **root, Node *x) {
    Node *y = x->left;
    x->left = y->right;
    if (y->right != NULL)
        y->right->parent = x;

    y->parent = x->parent;
    if (x->parent == NULL)
        *root = y;
    else if (x == x->parent->right)
        x->parent->right = y;
    else
        x->parent->left = y;

    y->right = x;
    x->parent = y;
}

void fixInsert(Node **root, Node *k) {
    while (k->parent != NULL && k->parent->color) {
        Node *u;
        if (k->parent == k->parent->parent->left) {
            u = k->parent->parent->right;

            if (u != NULL && u->color) { 
                k->parent->color = false;
                u->color = false;
                k->parent->parent->color = true;
                k = k->parent->parent;
            } else {
                if (k == k->parent->right) {
                    k = k->parent;
                    leftRotate(root, k);
                }
                k->parent->color = false;
                k->parent->parent->color = true;
                rightRotate(root, k->parent->parent);
            }
        } else {
            u = k->parent->parent->left;

            if (u != NULL && u->color) {
                k->parent->color = false;
                u->color = false;
                k->parent->parent->color = true;
                k = k->parent->parent;
            } else {
                if (k == k->parent->left) { 
                    k = k->parent;
                    rightRotate(root, k);
                }
                k->parent->color = false;
                k->parent->parent->color = true;
                leftRotate(root, k->parent->parent);
            }
        }
    }
    (*root)->color = false; 
}

void insert(Node **root, int key) {
    Node *node = newNode(key);
    Node *parent = NULL, *current = *root;

    while (current != NULL) {
        parent = current;
        if (key < current->data)
            current = current->left;
        else
            current = current->right;
    }

    node->parent = parent;
    if (parent == NULL)
        *root = node;
    else if (key < parent->data)
        parent->left = node;
    else
        parent->right = node;

    fixInsert(root, node);
}

void inorder(Node *root) {
    if (root != NULL) {
        inorder(root->left);
        printf("%d ", root->data);
        inorder(root->right);
    }
}

int main() {
    Node *root = NULL;

    insert(&root, 10);
    insert(&root, 20);
    insert(&root, 30);
    insert(&root, 15);

    printf("Inorder traversal: ");
    inorder(root);
    printf("\n");

    return 0;
}

File: Anirudh/rbt.py
class Node:
    def __init__(self, key):
        self.key = key
        self.color = "RED"  # New nodes are always red
        self.left = None
        self.right = None
        self.parent = None


class RedBlackTree:
    def __init__(self):
        self.NIL = Node(0)
        self.NIL.color = "BLACK"
        self.root = self.NIL

    def insert(self, key):
        new_node = Node(key)
        new_node.left = self.NIL
        new_node.right = self.NIL

        parent = None
        current = self.root

        # Standard BST insert
        while current != self.NIL:
            parent = current
            if new_node.key < current.key:
                current = current.left
            else:
                current = current.right

        new_node.parent = parent

        if parent is None:
            self.root = new_node
        elif new_node.key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Fix violations
        self.fix_insert(new_node)

    def fix_insert(self, node):
        while node != self.root and node.parent.color == "RED":
            if node.parent == node.parent.parent.left:
                uncle = node.parent.parent.right

                # Case 1: Uncle is red
                if uncle.color == "RED":
                    node.parent.color = "BLACK"
                    uncle.color = "BLACK"
                    node.parent.parent.color = "RED"
                    node = node.parent.parent

                else:
                    # Case 2: Node is right child
                    if node == node.parent.right:
                        node = node.parent
                        self.left_rotate(node)

                    # Case 3: Node is left child
                    node.parent.color = "BLACK"
                    node.parent.parent.color = "RED"
                    self.right_rotate(node.parent.parent)

            else:
                uncle = node.parent.parent.left

                # Mirror cases
                if uncle.color == "RED":
                    node.parent.color = "BLACK"
                    uncle.color = "BLACK"
                    node.parent.parent.color = "RED"
                    node = node.parent.parent

                else:
                    if node == node.parent.left:
                        node = node.parent
                        self.right_rotate(node)

                    node.parent.color = "BLACK"
                    node.parent.parent.color = "RED"
                    self.left_rotate(node.parent.parent)

        self.root.color = "BLACK"

    def left_rotate(self, x):
        y = x.right
        x.right = y.left

        if y.left != self.NIL:
            y.left.parent = x

        y.parent = x.parent

        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def right_rotate(self, x):
        y = x.left
        x.left = y.right

        if y.right != self.NIL:
            y.right.parent = x

        y.parent = x.parent

        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def inorder(self, node):
        if node != self.NIL:
            self.inorder(node.left)
            print(f"{node.key} ({node.color})", end=" ")
            self.inorder(node.right)


# Example usage
if __name__ == "__main__":
    rbt = RedBlackTree()
    values = [10, 20, 30, 15, 25, 5]

    for v in values:
        rbt.insert(v)

    print("Inorder traversal:")
    rbt.inorder(rbt.root)
